libexec/docker-import.sh

```diff
@@ -0,0 +1,227 @@
+if ! id=$(docker create "$dock" df -k -P /); then
+    message ERROR "docker create failed\n"
+    exit 1
+fi
+
+# Fixme: These should come from import.exec
+extra=${extra:-50}
+tarfactor=${tarfactor:-1.5}
+
+# Estimate the size of the running root.
+
+# Add an arbitrary $extra headroom for now in case of alterations or
+# as a possible fiddle factor for the filesystem, and hope $tarfactor
+# is a big enough factor by which to expand the tar stream for
+# singularity, else try again.  I.e. the initial attempt for the
+# singularity size is estimate*tarfactor+extra.
+
+# Starting docker with the cmd we gave it is presumably no good with
+# non-default entrypoint.  (We could actually use --entrypoint, though
+# it looks as if that has to be single word.)
+if [[ $entry = null ]]; then
+    size=$(docker start -a $id 2>/dev/null |
+           # Try to verify proper df output (checking the first line) to
+           # see if running df worked, and calculate the size.
+           awk -v extra="$extra" 'NR==1 && !/^Filesystem/ {exit 1}
+                                  NR==2 {print int($3/1024+extra)}')
+fi
+if [[ $? -ne 0 || $entry = null || ! $size =~ [0-9]+ ]]; then
+    # We have to fall back to guessing from the tar stream.
+    # Use awk as we may not have bc for the floating point calc.
+    if size=$(docker export $id | wc -c); then
+        size=$(awk "END {print int($tarfactor*$size/1024/1024+$extra)}" </dev/null)
+    else false
+    fi
+    if [[ $? -ne 0 ]]; then
+        message ERROR "Can't extract container size\n"
+        exit 1
+    fi
+fi
+
+message 1 "Creating $sing...\n"
+# redirect because -q doesn't work
+if ! singularity -q create -s $size "$sing" >/dev/null; then
+    message ERROR "failed: singularity create -s $size $sing\n"
+    exit 1
+fi
+
+## Export
+
+message 1 "Exporting/importing...\n"
+if ! docker export $id; then
+    message ERROR "docker export failed\n"
+    exit 1
+fi |
+  # If the import fails (presumably for lack of space), we expand the
+  # image and try again.  Sink stdout because of tar v and sink stderr
+  # in case of not-enough-space errors.
+  if ! singularity import "$sing" >/dev/null 2>&1; then
+      if ! singularity -q expand "$sing"; then
+          message ERROR "singularity expand failed\n"
+          exit 1
+      elif ! docker export $id | singularity import "$sing" >/dev/null; then
+          message ERROR "singularity import failed\n"
+          exit 1
+      fi
+  fi
+
+# Populate /singularity to reflect Docker semantics.  Quoting is a pain.
+# This generates a script stream to feed to singularity exec, avoiding a
+# scratch file, but perhaps the file would be better.
+# Note that the Docker semantics for the entrypoint and cmd aren't too
+# clear <https://docs.docker.com/engine/reference/builder/>.
+
+if ! cmd=$(docker inspect --format='{{json .Config.Cmd}}' "$dock"); then
+    message ERROR "docker inspect failed\n"
+    exit 1
+fi
+
+entry='["/bin/sh","-c"]'
```

Comment:
[Sorry this didn't get sent.  Yesterday must have been a while ago!]

You wrote: 

>> +entry='["/bin/sh","-c"]'
>
> Why do you statically define the entry point and command here?

Sorry, I checked in a version with that for testing.  I pushed a change
yesterday, but didn't get round to the pull request -- though I don't
understand how to refer to specific changes.